### PR TITLE
Improve TestPatch to make it much quicker when there are no conflicts and large files

### PR DESCRIPTION
**Pull request by zeripath**
Prior to this PR essentially the test patch code was git diff -p --binary mergebase base head | git apply --check --cached 

This is extremely slow if there are large binary files and the implementation dumped large temporary files on to the system.

We therefore propose a different option here:

* git read-tree -m to attempt the basic merge
* git status --porcelain can then be used to find the conflicts here.
* we can then limit the diff patch to only the conflicts - and we will drop the binary diff as the whitespace diffs matching is unlikely to be of help. 

Signed-off-by: Andrew Thornton <art27@cantab.net>

### CODE REVIEW

**Comment by zeripath** [comment]
This PR should save at least 12s in our tests.

**Comment by zeripath** [comment]
Ready

**Comment by a1012112796** [comment]
my proposal logic:
- 1. if the ``merge_base`` is the last commit on base branch, It will't have conflicts
- 2. get changed files in ``base_branch...merge_base`` and ``head_branch...merge_base``
- 3. If one file double changed in two list, It maybe have conflicts, should check it by ``git merge-file``

**Comment by zeripath** [comment]
@a1012112796 All of your changes are covered in 

```
git read-tree base
git read-tree -m --aggressive merge-base base tracking
git status --porcelain
```

and are much faster and better covered in that.

git read-tree -m --aggressive handles renames correctly amongst other things - your changes do not.

File: services/pull/patch.go [hunk 1]
```diff
@@ -98,139 +96,187 @@ func TestPatch(pr *models.PullRequest) error {
...
+	const empty = "error: unrecognised input"
+
+	const prefix = "error: patch failed:"
+	const errorPrefix = "error: "
+
+	conflictMap := map[string]bool{}
+
+	// 5f. Now scan the output from the command
+	scanner := bufio.NewScanner(stderrReader)
+	for scanner.Scan() {
```
Comment (a1012112796) [inline services/pull/patch.go]:
```suggestion
	// 5f. Now scan the output from the command
	// only list 10 conflicted files
	scanner := bufio.NewScanner(stderrReader)
	for len(conflictMap) < 10 && scanner.Scan() {
```

File: services/pull/patch.go [hunk 2]
```diff
@@ -98,139 +96,187 @@ func TestPatch(pr *models.PullRequest) error {
...
+		line := scanner.Text()
+		if strings.EqualFold(line, empty) && !conflict {
+			// ignore
+		} else if strings.HasPrefix(line, prefix) {
```
Comment (a1012112796) [inline services/pull/patch.go]:
```suggestion
		if strings.EqualFold(line, empty) && !conflict {
			// ignore
			continue
		}
		
		if strings.HasPrefix(line, prefix) {
```

File: services/pull/patch.go [hunk 3]
```diff
@@ -98,139 +96,187 @@ func TestPatch(pr *models.PullRequest) error {
...
+			conflict = true
+			filepath := strings.TrimSpace(strings.Split(line[len(prefix):], ":")[0])
+			conflictMap[filepath] = true
+		} else if strings.HasPrefix(line, errorPrefix) {
+			conflict = true
```
Comment (a1012112796) [inline services/pull/patch.go]:
```suggestion
			continue
		}
		
		if strings.HasPrefix(line, errorPrefix) {
			conflict = true
```

File: services/pull/patch.go [hunk 4]
```diff
@@ -98,139 +96,187 @@ func TestPatch(pr *models.PullRequest) error {
...
-						pr.ConflictedFiles = append(pr.ConflictedFiles, key)
 					}
+					break
 				}
+			}
+		}
+		// only list 10 conflicted files
+		if len(conflictMap) >= 10 {
+			break
+		}
```
Comment (a1012112796) [inline services/pull/patch.go]:
```suggestion
```

**Review by a1012112796 — COMMENTED** [review]
Some logic simplify about ``else`` :)

File: services/pull/patch.go [hunk 5]
```diff
@@ -98,139 +96,187 @@ func TestPatch(pr *models.PullRequest) error {
...
+			return
+		}
+		_ = stderrWriter.Close()
+	}()
+	const empty = "error: unrecognised input"
+
+	const prefix = "error: patch failed:"
+	const errorPrefix = "error: "
+
+	conflictMap := map[string]bool{}
```
Comment (a1012112796) [inline services/pull/patch.go]: why using map?

**Comment by a1012112796** [comment]
Now this function is a little long(190 lines) ..........
Maybe it's better to split it...

File: services/pull/patch.go (hunk 5 above)
Comment (zeripath) [reply services/pull/patch.go]: So unfortunately the same file can be listed multiple times.

**Comment by zeripath** [comment]
Have split it up and added a few more comments

**Review by silverwind — APPROVED** [review]

File: services/pull/patch.go [hunk 6]
```diff
@@ -98,139 +96,112 @@ func TestPatch(pr *models.PullRequest) error {
 	return nil
 }
 
+const conflictStatus = "DD AU UD UA DU AA UU "
```
Comment (6543) [inline services/pull/patch.go:100]:
```suggestion
// conflictStatus info can be found at https://git-scm.com/docs/git-status#_short_format
const conflictStatus = "DD AU UD UA DU AA UU "
```

Comment (6543) [reply services/pull/patch.go:100]: just some docs note ... if other will have same question ;)

**Comment by 6543** [comment]
this did not recognize a custom created conflict:

to reproduce:
 * create two branches from master (A & B)
 * add file `CONFL` with content `aaa\n` to branch A
 * add file `CONFL` with content `bbb\n` to branch B
 * create pull from A to master
 * create pull from B to master
 * merge B
 * look at A: still show "merge ..." button!

**Comment by zeripath** [comment]
Damn - I'll take a look. In the meantime let me mark this as WIP

**Comment by zeripath** [comment]
Replaced by #18004

**Pull request closed without merging**